File: apps/front/src/models/product.ts
export type Product<T extends ProductMetadata> = {
  id: string
  name: string
  description: string
  price: number
  category: ProductCategory
  soldCount: number
  metadata: T
  createdAt: Date
}

export enum ProductCategory {
  Upgrade = 'upgrade',
  NameColor = 'name-color',
  NamePrefix = 'name-prefix',
}

export type ProductMetadata = object

export type UpgradeProductMetadata = ProductMetadata & {
  action: string
}

export type NameColorProductMetadata = ProductMetadata & {
  colors: string[]
}

export type NamePrefixProductMetadata = ProductMetadata & {
  prefix: string
  namePrefixId: string
}

File: apps/api/internal/transport/http/responses/product.go
package responses

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type Product struct {
	ID          uuid.UUID       `json:"id"`
	Price       float64         `json:"price"`
	Name        string          `json:"name"`
	Description string          `json:"description"`
	Category    string          `json:"category"`
	Metadata    json.RawMessage `json:"metadata"`
	SoldCount   int64           `json:"soldCount"`
	CreatedAt   time.Time       `json:"createdAt"`
}
